File: HumanPlayer.cpp
#include <iostream>
#include <sstream>
#include "HumanPlayer.h"
#include "Grid.h"
#include "StringUtil.h"

using namespace std;

namespace {
    void outputInvalidInput() {
        cout << "Invalid input: 2 digits (1-" << Grid::SIZE << ") separated with a space must be given" << endl;
    }

    bool _checkAndSetValue(short value, short& result) {
        result = (short) (value - 1);
        return result >= 0 && result < Grid::SIZE;
    }
}

Turn HumanPlayer::makeTurn() {
    Turn turn;
    turn.playerId = id();
    while(!getInput(turn)) {}
    return turn;
}

bool HumanPlayer::getInput(Turn& turn) {
    string s;
    cout << "Player " << id() << ": ";
    getline(cin, s);

    string buf;
    stringstream ss(s);

    int numStr = 0;
    while (ss >> buf) {
        short inputValue;
        if (numStr == 2 || !StringUtil::isNumber(buf) || !_checkAndSetValue((short) stoi(buf), inputValue)) {
            outputInvalidInput();
            return false;
        }
        (&turn.row)[numStr] = inputValue;
        ++numStr;
    }
    if (numStr != 2) {
        outputInvalidInput();
        return false;
    }
    return true;
}

File: HumanPlayer.lua
require "Player"

HumanPlayer = Player:new()

function HumanPlayer:new(id)
    local o = Player:new(id)
    setmetatable(o, self)
    self.__index = self
    return o;
end

function HumanPlayer:makeTurn()
    local turn = Turn:new()
    turn.playerId = self:id();
    while(not self:getInput(turn)) do end
    print()
    return turn
end

function HumanPlayer:getInput(turn)
    io.write(string.format("Player %d: ", self:id()))
    io.flush()
    local input = io.read()

    local numStr = 0
    for i in string.gmatch(input, "%S+") do
        local inputValue = tonumber(i)
        if (numStr == 2 or inputValue == nil or not self:isInRange(inputValue)) then
            self:outputInvalidInput()
            return false
        end
        if (numStr == 0) then
            turn.row = inputValue - 1
        else
            turn.column = inputValue - 1
        end
        numStr = numStr + 1
    end

    if (numStr ~= 2) then
        self:outputInvalidInput()
        return false
    end

    return true
end

function HumanPlayer:isInRange(value)
    return value > 0 and value <= GRID_SIZE
end

function HumanPlayer:outputInvalidInput()
    print(string.format("Invalid input: 2 digits (1-%d) separated with a space must be given\n", GRID_SIZE))
end